### DISCUSSION TITLE
Added a BME280 weather sensor to the camera

### DISCUSSION THREAD

**Discussion by FBMinis** [jameszah/ESP32-CAM-Video-Telegram#50] [Show and tell]
I have a camera pointed at the horizon sending me photos periodically. I thought I'd add a weather sensor and later on a rain detector.

I used the BME280 sensor for pressure, humidity and temperature. The original Adafruit libraries have a variable that had to be renamed, as it conflicted with a variable inside `esp_camera.h`.

The modified Adafruit libraries can be found here: https://github.com/abish7643/ESP32Cam-I2CSensors/tree/master/lib

Connections:

![ESP32Cam - BME280](https://github.com/jameszah/ESP32-CAM-Video-Telegram/assets/36670323/99b1e47a-17bd-421e-a470-61fb8f455b1f)

Features:

- based on `ESP32-CAM-Video-Telegram_9.5`
- use /weather to get a message with current weather data (sensor is updated every 60 seconds);
- weather data will be displayed as a caption under a Timed Photo;

Code:

[ESP32-CAM-Video-Telegram_BME280.txt](https://github.com/jameszah/ESP32-CAM-Video-Telegram/files/12420310/ESP32-CAM-Video-Telegram_BME280.txt)

Thank you

**Comment by jameszah** [comment]
Another way to do that is to separate the adafruit stuff from the camera stuff, so the symbols to not step on each other.

https://github.com/jameszah/ESP32-CAM-RocketCam/blob/main/ESP32-CAM-Video-Recorder-junior-60x.4.7.srt.4a/sensors.cpp

**Reply by FBMinis** [reply]
So you just place the sensor stuff in a different .cpp (sensors.cpp) and then use the `extern `and `&` to pass the values into the main sketch by reference? Can you give me some keyworks so I can search about doing this sort of stuff?

**Reply by jameszah** [reply]
I take it you did not lose a year of your life in 2nd year software engineering. 😀
I think "variable scope in cpp" will give you lots of stuff.  
It does not come up that much in the tiny memory of a microcontroller, but giant software projects with a millions lines of code, the scope and namespace stuff keeps everything segregated with only "official" connections between the different blocks of code.

**Reply by FBMinis** [reply]
I learned a bit of Fortran and Matlab almost 20 years ago while studying Environmental Engineering and that was it! I will give it a shot this  evening. I only have a BMP280 here (not the BME) but it also presented the same conflict between variables, so I'll try to reproduce your approach. I will also try to add an ADS1115 ADC module to experiment reading battery voltage in my battery powered cameras, as there seems to be some trouble using the ADC of the ESP32 camera modules.

**Comment by FBMinis** [comment]
Well, the humidity reading began staying mostly stuck at 100%, which I discovered to be a common problem for many hobbyists. The temperature and pressure readings are in accordance with the weather station at the airfield nearby. 

I ordered one of these to replace the BME280: https://www.aliexpress.com/item/1005005581974032.html

(nice shop with inexpensive ESP32Cam and SR620 sensors too).

**Reply by jameszah** [reply]
You can imagine that tiny hole where the air enters the sensor could get dirty or clogged or never worked in tge first place.
